[SDK] Unit tests for TrainingClient APIs - get_job_pod_names and update_job

YosiElias opened the pull request:
<!--  Thanks for sending a pull request! Here are some tips for you:
1. If this is your first time, check our contributor guidelines: https://www.kubeflow.org/docs/about/contributing
2. To know more about Training Operator, check the developer guide:
    https://github.com/kubeflow/training-operator/blob/master/docs/development/developer_guide.md
3. If you want *faster* PR reviews, check how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
-->

**What this PR does / why we need it**:
New unit test for get_job_pod_names and update_job functions

**Which issue(s) this PR fixes** _(optional, in `Fixes #<issue number>, #<issue number>, ...` format, will close the issue(s) when PR gets merged)_:
Fixes #2161

**Checklist:**

- [ ] [Docs](https://www.kubeflow.org/docs/components/training/) included if any changes are user facing

YosiElias commented:
can someone pls review this PR? 
@andreyvelich @jinchihe @kuizhiqing

andreyvelich commented on sdk/python/kubeflow/training/api/training_client_test.py:
```diff
@@ -17,11 +18,23 @@
 import pytest
 
-LIST_RESPONSE = [{"metadata": {"name": "Dummy V1PodList"}}]
+LIST_RESPONSE = [
+    {"metadata": {"name": "Dummy V1PodList-1"}},
+    {"metadata": {"name": "Dummy V1PodList-2"}},
+]
```
Why do you want to change it ?

YosiElias replied:
Just to test and present the possibility of returning more than one name (I can remove this case if you think its unnecessary)

YosiElias replied:
@andreyvelich is it fine by you to keep this case?

andreyvelich replied:
I would suggest keep it as it is for old tests, since in that case we don't need to introduce SimpleNamespace from typing module.

andreyvelich commented on sdk/python/kubeflow/training/api/training_client_test.py:
```diff
@@ -250,14 +272,72 @@ def __init__(self, kind) -> None:
+        "valid flow with 2 pods",
+        {
+            "name": TEST_NAME,
+            "namespace": "2_pods",
```
Do you need to use different namespaces here ?

andreyvelich reviewed (commented):
Sorry for the late review @YosiElias!
Please take a look
/assign @droctothorpe @kubeflow/wg-training-leads @Electronic-Waste

andreyvelich commented:
Hi @YosiElias! Did you get a chance to check my comment here: https://github.com/kubeflow/training-operator/pull/2192#discussion_r1722079807 ?
I think after that we should be ready to merge this PR.

YosiElias commented:
Hi @andreyvelich! Sorry for the delay, I was busy with other things.
As you suggested I changed the 2 pod case (and dropped the use of SimpleNamespace) , if there are any additional comments I would love to hear them.
Thank you very much for your comments!

andreyvelich commented on sdk/python/kubeflow/training/api/training_client_test.py:
```diff
@@ -31,12 +33,28 @@ def create_namespaced_custom_object_response(*args, **kwargs):
                 raise multiprocessing.TimeoutError()
             if args[0] == "runtime":
                 raise Exception()
+            if args[0] == "mock_pod_obj":
+                pod_obj = Mock(metadata=Mock())
+                pod_obj.metadata.name = "Dummy V1PodList"
+                return Mock(items=[pod_obj])
+            if args[0] == "no_pods":
+                return Mock(items=[])
```
Do we want to make these namespace as constants and re-use it in other vars: `timeout, runtime, mock_pod_obj, and no_pods` ?

YosiElias replied:
I agree, I'll change it to constant. Thanks for the comment!

andreyvelich commented:
Overall, looks good, thank you for doing this @YosiElias!
Please fix the pre-commit, so we can merge it.
/assign @droctothorpe @kubeflow/wg-training-leads

andreyvelich reviewed (commented):
Thanks for the update @YosiElias!
/lgtm
/approve
